[Electron] Support click `Show Source` to open file in editor

jhen0409 opened the pull request:
This PR made click `Show Source` to open file in editor in Electron part, if we used [`babel-plugin-transform-react-jsx-source`](https://www.npmjs.com/package/babel-plugin-transform-react-jsx-source)):

![2017-03-19 22_37_24](https://cloud.githubusercontent.com/assets/3001525/24081771/a5562680-0cf4-11e7-9cbb-99fb35f76dcf.gif)

> Example for RN

* Add `launchEditor`, it's mirror from [react-native packager](https://github.com/facebook/react-native/blob/master/local-cli/server/util/launchEditor.js), have changes:
  - Remove `projectRoots`, we shouldn't get it here
  - Clear logs
* Add showComponentSource prop for open file in editor

suchipi commented on packages/react-devtools-core/src/backend.js:
```diff
@@ -15,6 +15,7 @@ type ConnectOptions = {
   port?: number,
   resolveRNStyle?: (style: number) => ?Object,
   isAppActive?: () => boolean,
+  openFileInEditor?: () => {},
```
This type doesn't look right; [openFileInEditor](https://github.com/facebook/react-native/blob/master/Libraries/Core/Devtools/openFileInEditor.js) has no return value. It should probably be `() => void` instead.

jhen0409 replied:
@suchipi you're right, thank you! 👍

suchipi commented on plugins/ReactNativeOpenFile/setupBackend.js:
```diff
@@ -0,0 +1,34 @@
+
+import type Bridge from '../../agent/Bridge';
+
+module.exports = function setupRNOpenFile(bridge: Bridge, openFileInEditor: () => {}) {
```
The type should be `() => void` here, too.

gaearon commented on plugins/ReactNativeOpenFile/setupBackend.js:
```diff
@@ -0,0 +1,34 @@
+    ) {
+      return;
+    }
+    var source = el._reactInternalInstance._currentElement._source;
```
This code is coupled to React internal representation (which is changing as we speak). Can you rewrite it so that it uses the “data” (see `getData()`) just like the frontend code? This way this won’t break when we switch to Fiber.

jhen0409 replied:
Got it. I rewrote it with `getData` and `getDataFiber`, I refer to [this implementation](https://github.com/facebook/react/blob/4b2eac3de7e1dbf5c2dd742fd9989974a83972cb/src/shared/utils/getComponentName.js) for check Fiber.

gaearon commented on plugins/ReactNativeOpenFile/setupBackend.js:
```diff
@@ -0,0 +1,41 @@
+      (prev, curr) => prev ? prev[curr] : null,
+      window
+    );
+    var inst = el._reactInternalInstance;
```
Sorry I wasn't clear, I didn't mean importing `getData` itself. I meant accessing the `node` data structure (that is generated by `getData` when renderer attaches). You are still accessing the "opaque nodes" (`el`) which is a problem (they're implementation details).

Agent has access to "data" in `this.elementData`. However I don't think it's meant to be a public field.

I would look at `showComponentSource` in `Panel` instead. This looks like related code (not sure if it's used though). Perhaps this is a better place to implement it, and somehow inject an implementation there?

jhen0409 replied:
> I would look at showComponentSource in Panel instead.

I tried `node.get('source')` just worked in `Panel`, obviously I did something unnecessarily. 😂

Also, I think it would be great if we just do [`launchEditor`](https://github.com/facebook/react-native/blob/master/local-cli/server/util/launchEditor.js) in Electron part, this will don't need go to RN process (call openFileInEditor) and support non-RN platforms, what do you think?

gaearon replied:
>I tried node.get('source') just worked in Panel, obviously I did something unnecessarily. 😂

Sounds awesome!

>I think it would be great if we just do launchEditor in Electron part, this will don't need go to RN process (call openFileInEditor) and support non-RN platforms, what do you think?

I don’t fully understand but if it’s better, go for it!

jhen0409 commented:
@gaearon I changed this PR as https://github.com/facebook/react-devtools/pull/588#discussion_r112542394.

* Add `launchEditor`, it's mirror from [react-native packager](https://github.com/facebook/react-native/blob/master/local-cli/server/util/launchEditor.js), have changes:
  - Remove `projectRoots`, we shouldn't get it here
  - Clear logs
* Add showComponentSource prop for open file in editor

Due to I removed `projectRoots` so I guess some editors cannot open file to the correct workspace, but currently I can sure it working on Atom.

gaearon commented:
This looks awesome. I love it. Thanks!

gaearon commented:
Please see also: #625.
